jinahub/indexers/query/vector/AnnoyIndexer/__init__.py

```diff
@@ -17,43 +20,90 @@ class AnnoyIndexer(BaseNumpyIndexer):
+        self.indexer = AnnoyIndex(self.num_dim, self.metric)
+        self.traverse_path = traverse_path
+        if os.path.exists(self.index_path) and os.path.exists(self.index_map_path):
+            self._load_index()
 
-    def build_advanced_index(self, vecs: 'np.ndarray'):
-        """Build an advanced index structure from a numpy array.
+    def _load_index(self):
+        self.indexer.load(self.index_path)
+        with open(self.index_map_path, 'r') as f:
+            self.id_docid_map = json.load(f)
 
-        :param vecs: numpy array containing the vectors to index
-        """
-        from annoy import AnnoyIndex
-        _index = AnnoyIndex(self.num_dim, self.metric)
-        for idx, v in enumerate(vecs):
-            _index.add_item(idx, v.astype(np.float32))
-        _index.build(self.n_trees)
-        return _index
-
-    def query(self, keys: 'np.ndarray', top_k: int, *args, **kwargs) -> Tuple['np.ndarray', 'np.ndarray']:
-        """Find the top-k vectors with smallest ``metric`` and return their ids in ascending order.
-        :param keys: numpy array containing vectors to search for
-        :param top_k: upper limit of responses for each search vector
-        """
-        all_idx = []
-        all_dist = []
-        for k in keys:
-            idx, dist = self.query_handler.get_nns_by_vector(k, top_k, self.search_k, include_distances=True)
-            all_idx.append(self._int2ext_id[self.valid_indices][idx])
-            all_dist.append(dist)
-        return np.array(all_idx), np.array(all_dist)
+    @requests(on='/delete')
+    def delete(self, **kwargs):
+        try:
+            os.remove(self.index_path)
+            os.remove(self.index_map_path)
+        except OSError as e:
+            if e.errno != errno.ENOENT:
+                raise
+
+    @requests(on='/index')
+    def index(self, docs: DocumentArray, **kwargs):
+        self.request_type = '/index'
+        if os.path.exists(self.index_path) or os.path.exists(self.index_map_path):
+            raise FileExistsError(
+                'Index already exist, please remove workspace and index again.'
+            )
+
+        chunks = docs.traverse_flat(self.traverse_path)
+
+        for idx, doc in enumerate(chunks):
+            self.id_docid_map[idx] = doc.id
+            self.indexer.add_item(idx, doc.embedding)
+
+    @requests(on='/search')
+    def search(self, docs: DocumentArray, **kwargs):
+        chunks = docs.traverse_flat(self.traverse_path)
+        for doc in chunks:
+            indices, dists = self.indexer.get_nns_by_vector(
+                doc.embedding, self.top_k, include_distances=True
+            )
+            for idx, dist in zip(indices, dists):
+                match = Document(id=self.id_docid_map[str(idx)])
+                match.score.value = 1 / (1 + dist)
+                doc.matches.append(match)
+
+    def close(self):
+        if self.request_type == '/index':
```

Comment:
This pattern should not be here. `self.request_type` should never be kept or such state mantained

Reply:
Sure, fixed with the suggestion in slack